core/services/chainlink/application.go

```diff
@@ -798,9 +798,12 @@ func (app *ChainlinkApplication) ReplayFromBlock(chainID *big.Int, number uint64
 	}
 	chain.LogBroadcaster().ReplayFromBlock(int64(number), forceBroadcast)
 	if app.Config.FeatureLogPoller() {
-		if err := chain.LogPoller().Replay(context.Background(), int64(number)); err != nil {
-			return err
-		}
+		go func() {
+			err := chain.LogPoller().Replay(context.Background(), int64(number))
```

Comment:
Why do we use a background context here?

Reply:
I have a version of this PR which uses the request context, but after comparing the pros and cons of the two I decided `context.Background()` seems like the better choice here for several reasons.  Using the request context is messier, because it automatically gets cancelled as soon as this function returns.  And because we're calling `Replay()` in a separate goroutine that means it can be cancelled even before `Replay()` is called.  The only fix for that would be to have a timer that waits for a few seconds before returning, in the hope that the replay signal gets received before the timeout expires.  But there's not really any upper limit to how long we might need to wait... for example, if another Replay is already in progress it might have to wait several minutes.  Seems better to just ignore the request context and return immediately, firing it off in the background.

On the pro side of using the request, the reason I initially preferred that was because it would allow us to return the error code, and make that into a message that gets sent back to the user, informing them whether the replay is now complete, in progress, or if there was some other error.  That seems nice, but less so when you realize that the user-facing command is not specific to log poller, it just looks like this `chainlink blocks replay --block-number BLOCK_NUMBER`.  Internally, it sends a fully async signal to log broadcaster and to log poller.  Log broadcaster already uses background context.  So that's another point in favor of using background context for log poller:  it would be strange to have it wait for log poller to give some kind of preliminary status code, but not wait for log broadcaster.  Then difficult to explain to the user that this is only partial information that applies to one and not to the other.  Perhaps at some point we can change both of them, but for now this seems better.

Reply:
I'm concerned that an explanation requires so much knowledge about things out of scope for a reader of this code.
Is there something succinct we can put in a code comment?

Otherwise, this seems like a code smell, and I don't completely follow the rationale, since we are control of this API - not at its mercy.

Reply:
I'm concerned that an explanation requires so much knowledge about things out of scope for a reader of this code.
Is there something succinct we can put in a code comment?

Otherwise, this seems like a code smell, and I don't completely follow the rationale, since we are control of this API - not at its mercy.

Reply:
This is why I suggested a separate method:
```go
func (lp) ReplayAsync() {
	go func() {
		lp.Replay(lp.ctx)
	}
}
Then you spawn the goroutine alongside the background context that you want to use. Much of the complexity disappears, and you don't have to chain together mismatched APIs.

Reply:
Yeah, I thought about using the server context.  My only hesitation was that it goes against the idea of trying to keep a clean separation between the two contexts, and only use lp.ctx in lp.run().  I think this is not as bad as passing the other context over a channel though.  And we already do this in many other places in the codebase.

One thing I had thought of doing was removing lp.ctx from the struct and just making it a local variable in lp.run().  The only reason I didn't was because some of the tests need to inject a different context there, and/or wait on lp.ctx.Done() for sychronization.  But if it's going to be there anyway, it seems a bit silly to not make use of it.

context.Background() should be just as good as lp.ctx in this case, so I don't think it makes much difference.  But I agree that it's easier to look at the code and know that everything is safe if we use lp.ctx... and we've been in general trying to eliminate use of context.Background()

Reply:
Actually, there is a bit of a race here, because this intermediate goroutine does not block shutdown (even if you passed the server context). If shutdown is triggered right after this spawns, the whole system can clean itself up and return without this goroutine ever being executed. IIUC it would still be benign for it to execute afterwards, but that is just the current state and not enforced going forward. If we spawn the goroutine inside of the server code, then more important than having `lp.ctx` on hand, we could also register to block close until returning. (Unfortunately it looks like it is not set up with a `WaitGroup` so not completely trivial :thinking: )